src/trie_tree.rs

```diff
@@ -0,0 +1,456 @@
+            } else if last_height != 0 {
+                last_height -= 1;
+            } else {
+                // do nothing with a zero insertion
+                break;
+            }
+        }
+
+        for height in last_height..=core::u8::MAX {
+            // update tree hash from insert pos to top
+            let node_key = key.parent_path(height);
+            let branch_key = BranchKey::new(height, node_key);
+
+            let new_merge = if let Some(branch) = self.store.get_branch(&branch_key)? {
+                merge::<H>(height, &node_key, &branch.left, &branch.right)
+            } else {
+                MergeValue::zero()
+            };
+            if height == core::u8::MAX {
+                // updating root
+                self.root = new_merge.hash::<H>();
+            } else {
+                // updates parent branch
+                let parent_key = key.parent_path(height + 1);
+                let parent_branch_key = BranchKey::new(height + 1, parent_key);
+                if new_merge.is_shortcut() {
+                    // move up
+                    self.store.remove_branch(&branch_key)?;
+                }
+                if let Some(parent_branch) = self.store.get_branch(&parent_branch_key)? {
+                    let (left, right) = if key.is_right(height + 1) {
+                        (parent_branch.left, new_merge)
+                    } else {
+                        (new_merge, parent_branch.right)
+                    };
+                    if left.is_zero() && right.is_zero() {
+                        self.store.remove_branch(&parent_branch_key)?;
+                    } else {
+                        let new_parent_branch = BranchNode { left, right };
+                        self.store
+                            .insert_branch(parent_branch_key, new_parent_branch)?;
+                    }
+                } else if !new_merge.is_zero() {
+                    let new_parent_branch = if key.is_right(height + 1) {
+                        BranchNode {
+                            left: MergeValue::zero(),
+                            right: new_merge,
+                        }
+                    } else {
+                        BranchNode {
+                            left: new_merge,
+                            right: MergeValue::zero(),
+                        }
+                    };
+                    self.store
+                        .insert_branch(parent_branch_key, new_parent_branch)?;
+                }
+            }
+        }
+
+        Ok(&self.root)
+    }
+
+    /// Update multiple leaves at once
+    pub fn update_all(&mut self, mut leaves: Vec<(H256, V)>) -> Result<&H256> {
+        // Dedup(only keep the last of each key) and sort leaves
+        leaves.reverse();
+        leaves.sort_by_key(|(a, _)| *a);
+        leaves.dedup_by_key(|(a, _)| *a);
+
+        for (key, value) in leaves {
+            self.update(key, value)?;
+        }
```

Comment:
@quake 
Do we still need an optimized `update_all()` function?

Reply:
I don't think it's necessary, because with this PR algorithm, the optimization by `update_all` is very limited (around 1%)